solarforecastarbiter/reference_forecasts/main.py

```diff
@@ -465,3 +465,143 @@ def make_latest_nwp_forecasts(token, run_time, issue_buffer, base_url=None):
+
+    Returns
+    -------
+    generator of (Forecast, Observation, next_issue_time, index)
+
+    """
+    user_info = session.get_user_info()
+    observation_dict = {obs.observation_id: obs for obs in observations}
+    for fx in forecasts:
+        if not _is_reference_persistence_forecast(fx.extra_parameters):
+            logger.debug(
+                'Forecast %s is not labeled as a reference persistence forecast',  # NOQA
+                fx.forecast_id)
+            continue
+
+        if not fx.provider == user_info['organization']:
+            logger.debug(
+                "Forecast %s is not in user's organization",
+                fx.forecast_id)
+            continue
+
+        try:
+            extra_parameters = json.loads(fx.extra_parameters)
+        except json.JSONDecodeError:
+            logger.warning(
+                'Failed to decode extra_parameters for %s: %s as JSON',
+                fx.name, fx.forecast_id)
+            continue
+
+        try:
+            observation_id = extra_parameters['observation_id']
+        except KeyError:
+            logger.error(
+                ('Forecast, %s: %s, has no observation_id to base forecasts'
+                 ' off of. Cannot make forecast.'),
+                fx.name, fx.forecast_id)
+            continue
+        if observation_id not in observation_dict:
+            logger.error(
+                ('Observation %s not in set of given observations.'
+                 ' Cannot generate persistence forecast for %s:%s.'),
+                observation_id, fx.name, fx.forecast_id)
+            continue
+        observation = observation_dict[observation_id]
+
+        index = extra_parameters.get('index_persistence', False)
+        obs_mint, obs_maxt = session.get_observation_time_range(observation_id)
+        if pd.isna(obs_maxt):  # no observations to use anyway
+            logger.info(
+                'No observation values to use for %s:%s from observation %s',
+                fx.name, fx.forecast_id, observation_id)
+            continue
+
+        fx_mint, fx_maxt = session.get_forecast_time_range(fx.forecast_id)
+        # find the next issue time for the forecast based on the last value
+        # in the forecast series
+        if pd.isna(fx_maxt):
+            # if there is no forecast yet, go back a bit from the last
+            # observation. Don't use the start of observations, since it
+            # could really stress the workers if we have a few years of
+            # data before deciding to make a persistence fx
+            next_issue_time = utils.get_next_issue_time(
+                fx, obs_maxt - fx.run_length)
+        else:
+            next_issue_time = utils.find_next_issue_time_from_last_forecast(
+                fx, fx_maxt)
+
+        # now find all the run times that can be made based on the
+        # last observation timestamp
+        while next_issue_time <= max_run_time:
```

Comment:
Can this be a for loop over a DatetimeIndex? It's not immediately obvious that the while loop is guaranteed to exit.

Reply:
hmm, we should rely on ``utils.get_next_issue_time``/``utils.get_issue_times`` so we would first need a similar loop to construct the issue times. The nice thing here is that it is a generator instead of building a list and looping through it. I think if one trusts that ``util.get_next_issue_time`` starting from the current issue time (+ns) always returns a greater timestamp (it does) then the loop will exit. 

Basically, I'm confident the loop will exit, and the workers running the jobs have a finite lifetime per job, so we'll be able to spot issues.

Reply:
works for me